File: models/spacesim_models/substrate/modifiers.py
"""
spacesim_models.substrate.modifiers
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Maps second-order civilizational propensities to economic behavior modifiers.

BehaviorModifiers are derived from propensities and influence how a population
interacts with markets — how much they hoard, how sensitive prices are to
demand shocks, how much speculation inflates prices, etc.

This is the bridge between the biological/psychological model and the
market mechanics. The same basic needs can produce radically different
market dynamics depending on the species' propensity profile.
"""

from __future__ import annotations

from dataclasses import dataclass


@dataclass
class BehaviorModifiers:
    """
    Economic behavior modifiers derived from species propensities.

    demand_amplifier    : scales demand signal [1.0, 1.5]
                          driven by loss_aversion_premium — fearful populations
                          demand more aggressively than their satiation suggests

    supply_hoarding     : fraction of supply withheld from market [0.0, 0.4]
                          driven by stratification_tendency — elite capture
                          of supply reduces effective market availability

    price_sensitivity   : multiplier on price_elasticity [0.5, 2.0]
                          driven by volatility_under_stress — unstable societies
                          react sharply to imbalances, amplifying price swings

    cooperation_discount: price discount applied when demand is low [0.0, 0.3]
                          driven by cooperative_radius — cooperative societies
                          use mutual aid to dampen deflation in slack markets

    speculation_premium : extra premium added when demand is high [0.0, 0.15]
                          driven by short_termism — short-horizon societies
                          generate speculative bubbles above clearing price
    """
    demand_amplifier:     float = 1.25
    supply_hoarding:      float = 0.20
    price_sensitivity:    float = 1.25
    cooperation_discount: float = 0.15
    speculation_premium:  float = 0.075


def propensities_to_modifiers(props: dict[str, float]) -> BehaviorModifiers:
    """
    Derive BehaviorModifiers from a propensity dict.

    props should be the dict returned by compute_propensities_from_toml().
    All propensity values are in [0.0, 1.0].
    """
    return BehaviorModifiers(
        demand_amplifier     = 1.0  + 0.5  * props.get("loss_aversion_premium",      0.5),
        supply_hoarding      = 0.4  *        props.get("stratification_tendency",     0.5),
        price_sensitivity    = 0.5  + 1.5  * props.get("volatility_under_stress",    0.5),
        cooperation_discount = 0.3  *        props.get("cooperative_radius",          0.5),
        speculation_premium  = 0.15 *        props.get("short_termism",               0.5),
    )

File: sim/types/src/modifiers.rs
//! Economic behavior modifiers derived from species propensities.
//!
//! These translate the civilizational fingerprint (propensities) into
//! concrete adjustments on market clearing mechanics.

use serde::{Deserialize, Serialize};
use std::collections::HashMap;

/// Market behavior modifiers for a species.
///
/// Each field adjusts one dimension of market mechanics. All values are
/// derived from propensities via linear mappings; the ranges are calibrated
/// so that mid-range propensities (0.5) produce neutral modifier values.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BehaviorModifiers {
    /// Multiplier on demand signal [1.0, 1.5].
    /// Fearful populations (high loss_aversion_premium) demand more than their
    /// satiation alone would suggest — hoarding impulse amplifies apparent need.
    pub demand_amplifier: f64,

    /// Fraction of base supply withheld from the open market [0.0, 0.4].
    /// Stratified societies see elite supply capture reducing market availability.
    pub supply_hoarding: f64,

    /// Multiplier on price_elasticity [0.5, 2.0].
    /// Volatile societies (high volatility_under_stress) over-react to imbalances,
    /// producing larger price swings per unit of demand-supply gap.
    pub price_sensitivity: f64,

    /// Price discount applied in slack markets (demand < 1.0) [0.0, 0.3].
    /// Cooperative societies use mutual aid to dampen price drops, providing
    /// a floor that prevents destabilizing deflation spirals.
    pub cooperation_discount: f64,

    /// Speculative premium added when demand is high [0.0, 0.15].
    /// Short-termism drives speculative buying above clearing price,
    /// inflating bubbles during scarcity events.
    pub speculation_premium: f64,
}

impl BehaviorModifiers {
    /// Derive modifiers from the propensity dict returned by Python's api.py.
    pub fn from_propensity_map(props: &HashMap<String, f64>) -> Self {
        let get = |key: &str| props.get(key).copied().unwrap_or(0.5);
        BehaviorModifiers {
            demand_amplifier:     1.0  + 0.5  * get("loss_aversion_premium"),
            supply_hoarding:      0.4  *        get("stratification_tendency"),
            price_sensitivity:    0.5  + 1.5  * get("volatility_under_stress"),
            cooperation_discount: 0.3  *        get("cooperative_radius"),
            speculation_premium:  0.15 *        get("short_termism"),
        }
    }

    /// Neutral modifiers — midpoint values, used when no species is loaded.
    pub fn neutral() -> Self {
        BehaviorModifiers {
            demand_amplifier:     1.25,
            supply_hoarding:      0.20,
            price_sensitivity:    1.25,
            cooperation_discount: 0.15,
            speculation_premium:  0.075,
        }
    }
}
